src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -217,45 +215,58 @@ internal static bool ResolveOperationFromList(string identifier, IEnumerable<str
                     throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
                 }
 
-                matchingOperation = candidateMatchingOperations.Single();
+                matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
                 return true;
             }
 
-            if (foundActionsWhenLookingForFunctions.Count > 0)
+            if (foundActionsWhenLookingForFunctions != null && foundActionsWhenLookingForFunctions.Count > 0)
             {
                 throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
             }
 
             // If more than one overload matches, try to select based on optional parameters
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
-                candidateMatchingOperations = candidateMatchingOperations.FindBestOverloadBasedOnParameters(parameterNames);
+                candidatesMatchingOperations = candidatesMatchingOperations.FindBestOverloadBasedOnParameterCount(parameterNames.Count);
             }
 
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
                 throw new ODataException(ODataErrorStrings.FunctionOverloadResolver_NoSingleMatchFound(identifier, string.Join(",", parameterNames.ToArray())));
             }
 
-            matchingOperation = candidateMatchingOperations.SingleOrDefault();
+            matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
             return matchingOperation != null;
         }
 
         internal static bool HasAny<T>(this IEnumerable<T> enumerable) where T : class
         {
-            IList<T> list = enumerable as IList<T>;
-            if (list != null)
+            if (enumerable != null && enumerable.GetEnumerator().MoveNext())
             {
-                return list.Count > 0;
+                return true;
             }
 
-            T[] array = enumerable as T[];
-            if (array != null)
+            return false;
+        }
+
+        internal static bool CountGreaterThan<T>(this IEnumerable<T> enumerable, int count) where T : class

```

Comment:
move as private.

Reply:
not marked as private.

---
In reply to: [370898092](https://github.com/OData/odata.net/pull/1602#discussion_r370898092) [](ancestors = 370898092)

Reply:
I changed CountGreater than to private but ended up removing that. This is a shared function that gets used in SelectPathSegmentTokenBinder, therefore keeping as internal. 

---
In reply to: [371615196](https://github.com/OData/odata.net/pull/1602#discussion_r371615196) [](ancestors = 371615196,370898092)